### DISCUSSION TITLE
Why 5.x-dev faster than 4.x?

### DISCUSSION THREAD

**Discussion by cayolblake** [walkor/Workerman#728] [General]
Hi @walkor 

Why is 5.x-dev giving better performance by benchmarks ranging %10-%15?

Also, is it properly up to date like the 4.x tree?

**Comment by walkor** [comment]
5.x-dev has done some optimization in protocol parsing. If these optimizations prove feasible on 5.x-dev (it will take some time to verify), I will implement it on 4.x.

**Comment by cayolblake** [comment]
Well, in benchmarking I can verify it provides at least %20 improvements in amount of requests per second while maintaining low memory usage. This is for tests of 1 worker to 4 workers.

**Comment by cayolblake** [comment]
@walkor 

Here's the difference on a single machine, showing 5.x-dev is %30 faster than 4.0.30.

Seeing 4.x is more cared about and up to date, it would be great to see it get the same performance enhancements.

####  "^5.x-dev"
```
$ wrk -c1000 -t1 -d5 "http://127.0.0.1:2345/"

Running 5s test @ http://127.0.0.1:2345/
  1 threads and 1000 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    44.78ms  183.11ms   1.87s    94.81%
    Req/Sec   183.38k    17.34k  218.72k    70.00%
  910481 requests in 5.02s, 114.62MB read
Requests/sec: 181338.03
Transfer/sec:     22.83MB
```

#### "^4.0.30"
```
$ wrk -c1000 -t1 -d5 "http://127.0.0.1:2345/"

Running 5s test @ http://127.0.0.1:2345/
  1 threads and 1000 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    46.03ms  189.14ms   2.00s    95.23%
    Req/Sec   127.01k    21.88k  176.05k    60.00%
  630184 requests in 5.02s, 79.33MB read
  Socket errors: connect 0, read 0, write 0, timeout 90
Requests/sec: 125495.24
Transfer/sec:     15.80MB
```

**Reply by cayolblake** [reply]
Hi @walkor 

Any soon plans to port there enhancements to 4.x?

When starting a new project, is it recommended to use 5.x?

**Reply by walkor** [reply]
We need to observe whether this performance optimization is stable enough on 5.x, which will take a long time. If you need these optimizations, consider using 5.x-dev.

**Comment by cayolblake** [comment]
So if I want stability I use 4.x and if I want to performance I use 5.x? 🤔

**Reply by walkor** [reply]
I think the performance of 4.x is good enough. There is no need to pursue performance excessively.
